### PR TITLE
[Consolidated] Rewrote the dispatcher reconciler

### LINKED ISSUE #245: [Consolidated] Dispatcher should implement Finalizer to properly clean up resources
### Background
Currently the dispatcher has a bunch of caches that it populates when reconciling a channel, those caches include the Kafka Sarama consumers. However, those caches don't get updated when the channel or subscriptions gets deleted. 

### Proposed Changes
1. Implement FinalizeKind and ObserveFinalizedKind
1. Cleanup all the caches in the dispatcher accordingly
    1. hostToChannelMap
    1. channelSubscription
    1. subsConsumerGroups
    1. subscriptions

**Related**
See discussion: https://github.com/knative-sandbox/eventing-kafka/pull/182#discussion_r533362536

 **Expected behavior**

Resources are properly cleaned up.

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

Fix #245

## Proposed Changes

- :broom: Now we reconcile one channel at the time, more than doing a global resync every time
- :broom: Removed some bad concurrency patterns on the hot path
- :broom: Implemented CleanupChannel in dispatcher

### CODE REVIEW

File: pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go
Comment (markusthoemmes) [inline pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:165]: FWIW, this is "wrong", see https://github.com/knative/pkg/issues/2021

Comment (slinkydeveloper) [reply pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:165]: Isn't adding both fine? The `CleanupChannel` method is idempotent. If it's going to cause issues, I can remove it

Comment (slinkydeveloper) [reply pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go]: I've removed both FinalizeKind and ObserveFinalizeKind and replaced with a specific handler on the informer

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (devguyio) [inline pkg/channel/consolidated/dispatcher/dispatcher.go:265]: nit: I started thinking that relying on defer to guarantee unlock takes place even in panics is more robust? but I still don't have a super strong opinion about it. so I'd use an anonymous function for short scope locking here. wdyt?

Comment (slinkydeveloper) [reply pkg/channel/consolidated/dispatcher/dispatcher.go:265]: but this lock shouldn't be here at all :smile: set/get hostToChannelMap already uses a CAS pointer

Comment (devguyio) [reply pkg/channel/consolidated/dispatcher/dispatcher.go:265]: yeah....we don't speak about that though :smile: just saying in case another one tries to acquire it and this cleanup goroutine panics, the lock will stay locked IIUC.

**Review by devguyio — COMMENTED** [review]
Looks good to me, lmk what you think about using defer for unlock, it's a nit and otherwise I'm fine approving it.

**Comment by devguyio** [comment]
@slinkydeveloper seems like the race I was talking about is happening. This will show up when concurrent tests are running in parallel

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (matzew) [inline pkg/channel/consolidated/dispatcher/dispatcher.go:244]: have we gh issue for track this, as an investigtation?

File: pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:134]: what's the fix (or issue for the `FinalizeKind`) ?

Comment (matzew) [reply pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:134]: ah, I see: https://github.com/knative/pkg/issues/2021

thanks

**Comment by slinkydeveloper** [comment]
> seems like the race I was talking about is happening. This will show up when concurrent tests are running in parallel. You see in the test logs that they can't the channels they expect and I believe that's cause a previous "unsubscribe" deletes that channel somehow. Irrelevant channels disappear from the cache.

I guess the finalizer kind surface it right? What do you want me to do?

File: pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go
Comment (slinkydeveloper) [reply pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:134]: https://github.com/knative-sandbox/eventing-kafka/pull/509#discussion_r608467643

File: pkg/channel/consolidated/dispatcher/dispatcher.go
Comment (slinkydeveloper) [reply pkg/channel/consolidated/dispatcher/dispatcher.go:244]: Nope, should I create one?

**Comment by devguyio** [comment]
> I guess the finalizer kind surface it right? What do you want me to do? May I help somehow? Should I go forward with this PR or not?

yeah, seems very nasty to me but I believe we need to fix it. Are you game to dig this one out? Maybe give it a shot and see if you'll be able to figure out where the race is?

**Comment by slinkydeveloper** [comment]
@devguyio Ok will try

**Comment by devguyio** [comment]
/assign
@slinkydeveloper lmk when I can review, I took an initial look

**Comment by slinkydeveloper** [comment]
@devguyio You can start looking at it :smile: In my ideal world, e2e tests **should** pass :laughing:

**Comment by aliok** [comment]
I had a look and it seems fine.

I am working on KafkaChannel dispatcher since some time but I cannot review this PR properly. Pretty much everything changed.

There are fundamental changes but I trust your judgement. If the tests pass, I am fine with the PR.

**Comment by slinkydeveloper** [comment]
> I am working on KafkaChannel dispatcher since some time but I cannot review this PR properly. Pretty much everything changed.

I know, I'm sorry if I was drastical, but I saw no alternative to properly fix the whole dispatcher process which was racy and was leaving garbage...

**Comment by aliok** [comment]
> I know, I'm sorry if I was drastical, but I saw no alternative to properly fix the whole dispatcher process which was racy and was leaving garbage...

As I wrote earlier, I trust your judgement. If the tests pass, I am fine with the PR.

Thanks!

**Comment by aliok** [comment]
LGTM, leaving the approval to @devguyio

File: pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:158]: `ObserveKind` ?  it's in the `ReconcileKind`:
https://github.com/knative-sandbox/eventing-kafka/blob/a23efea868056e51a111b4f79f265c769b136fae/pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go#L157-L160

File: pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go
Comment (matzew) [inline pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:195]: newChannelConfigFromKafkaChannel ?

Comment (slinkydeveloper) [reply pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go]: renamed, what about now?

File: pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go
Comment (devguyio) [inline pkg/channel/consolidated/reconciler/dispatcher/kafkachannel.go:194]: typo, `newConfigFromKafkaChannel(S)` extra S

**Comment by matzew** [comment]
/lgtm
leaving approval for others

**Comment by matzew** [comment]
/hold
until other give approval

**Review by devguyio — APPROVED** [review]
Nice work :+1: I like how cleaner the new reconciliation logic is. I'll be working on respecting the subscription generation once this one is in :+1: 

/lgtm
/unhold

**Comment by matzew** [comment]
@slinkydeveloper there is a conflic w/ the 0.20 branch /cc @devguyio

**Comment by slinkydeveloper** [comment]
I will manually cherry pick for 0.21 and 0.20

**Pull request merged**
